RInAppMessaging/Classes/CampaignDispatcher.swift

```diff
@@ -58,41 +61,47 @@ internal class CampaignDispatcher: CampaignDispatcherType {
                 return
             }
 
-            var campaign = self.queuedCampaigns.removeFirst()
-
+            let campaign = self.queuedCampaigns.removeFirst()
             let permissionResponse = self.permissionService.checkPermission(forCampaign: campaign.data)
             if permissionResponse.performPing {
                 self.delegate?.performPing()
             }
+
             guard campaign.data.isTest || permissionResponse.display else {
                 self.dispatchNext()
                 return
             }
 
-            if let updatedCampaign = self.campaignRepository.decrementImpressionsLeftInCampaign(campaign) {
-                campaign = updatedCampaign
-            } else {
-                CommonUtility.debugPrint("""
-                    Error: Campaign (\(campaign.id)) does not exist in the repository anymore (race condition?). Proceeding with old data...
-                    """)
-            }
-
             let campaignTitle = campaign.data.messagePayload.title
             self.router.displayCampaign(campaign, confirmation: {
                 let contexts = campaign.contexts
                 guard let delegate = self.delegate, !contexts.isEmpty, !campaign.data.isTest else {
+                    self.campaignRepository.decrementImpressionsLeftInCampaign(campaign)
                     return true
                 }
-
-                return delegate.shouldShowCampaignMessage(title: campaignTitle,
-                                                          contexts: contexts)
-            }(), completion: { [weak self] in
-                guard let self = self else {
-                    return
+                let shouldShow = delegate.shouldShowCampaignMessage(title: campaignTitle,
+                                                                    contexts: contexts)
+                if shouldShow {
+                    self.campaignRepository.decrementImpressionsLeftInCampaign(campaign)
+                }
+                return shouldShow
+
+            }(), completion: { cancelled in
+                self.dispatchQueue.async(flags: .barrier) {
+
+                    guard !self.queuedCampaigns.isEmpty else {
+                        self.isDispatching = false
+                        return
+                    }
+                    if cancelled {
+                        self.dispatchNext()
+                    } else {
+                        WorkScheduler.scheduleTask(
+                            milliseconds: self.delayBeforeNextMessage(for: campaign.data),
```

Comment:
The minimum delay is 0?

Reply:
There is no delay if a message wasn't displayed (cancelled). The default delay is set in the backend and it was always 3s as long as I observed it for the first time.

Reply:
And the value is taken from `campaign.data`